src/delegate.rs

```diff
@@ -158,45 +156,46 @@ impl<T: LanguageServer> Delegate<T> {
 
         (delegate, messages)
     }
+}
 
-    fn delegate_notification<N, F>(&self, params: Params, delegate: F)
-    where
-        N: Notification,
-        N::Params: DeserializeOwned,
-        F: Fn(&Printer, N::Params),
-    {
-        if self.initialized.load(Ordering::SeqCst) {
-            match params.parse::<N::Params>() {
-                Ok(params) => delegate(&self.printer, params),
-                Err(err) => error!("invalid parameters for `{}`: {:?}", N::METHOD, err),
+macro_rules! delegate_request {
+    ($name:ident as $request:ty) => {
```

Comment:
I feel like using `as` here could be a bit misleading, what do you think of changing it to to something else like `->`
```suggestion
    ($name:ident -> $request:ty) => {
```
With calls looking like
```rust
delegate_request!(completion -> Completion);
```
To me this indicates more clearly that `Completion` has something to do with the return type.

Reply:
Nice idea! ❤️ I agree it makes it a bit more readable.